libs/@local/hashql/eval/src/graph/read/filter.rs

```diff
@@ -0,0 +1,158 @@
+                    }
+                    BinOpKind::Or => {
+                        let (left, right) = (
+                            self.compile_filter(context, left),
+                            self.compile_filter(context, right),
+                        );
+
+                        return Ok(Filter::Any(vec![left?, right?]));
+                    }
+                    BinOpKind::Eq => |left, right| Filter::Equal(Some(left), Some(right)),
+                    BinOpKind::Ne => |left, right| Filter::NotEqual(Some(left), Some(right)),
+                    BinOpKind::Lt => Filter::Less,
+                    BinOpKind::Lte => Filter::LessOrEqual,
+                    BinOpKind::Gt => Filter::Greater,
+                    BinOpKind::Gte => Filter::GreaterOrEqual,
+                };
+
+                let (left, right) = (
+                    self.compile_filter_expr(context, left)
+                        .and_then(|expr| expr.finish(context, &mut self.diagnostics)),
+                    self.compile_filter_expr(context, right)
+                        .and_then(|expr| expr.finish(context, &mut self.diagnostics)),
+                );
+
+                Ok(func(left?, right?))
+            }
+            // Unary are currently not supported, so we can skip them
+            NodeKind::Operation(Operation {
+                span: _,
+                kind:
+                    OperationKind::Type(TypeOperation {
+                        span: _,
+                        kind:
+                            TypeOperationKind::Assertion(TypeAssertion {
+                                span: _,
+                                value,
+                                r#type: _,
+                                force: _,
+                            }),
+                    }),
+            }) => self.compile_filter(context, value),
+            // If we came to this match arm using these nodes, then that means that the filter
+            // must've evaluated to a boolean expression. Therefore we can just check:
+            // did the expression evaluate to true?
+            NodeKind::Operation(Operation {
+                span: _,
+                kind:
+                    OperationKind::Type(TypeOperation {
+                        span: _,
+                        kind: TypeOperationKind::Constructor(..),
+                    }),
+            })
+            | NodeKind::Input(_)
+            | NodeKind::Data(_)
+            | NodeKind::Access(_)
+            | NodeKind::Call(_)
+            | NodeKind::Closure(_)
+            | NodeKind::Graph(_) => Ok(Filter::Equal(
+                Some(
+                    self.compile_filter_expr(context, node)?
+                        .finish(context, &mut self.diagnostics)?,
+                ),
+                Some(FilterExpression::Parameter {
+                    parameter: Parameter::Boolean(true),
+                    convert: None,
+                }),
+            )),
```

Comment:
Doesn't this recurse infinitely?

Reply:
Ah, I guess it's the `.finish()` which is doing the trick?

Reply:
These are two different functions. `compile_filter` is for the top level (where `&&` and `||` are allowed) while `compile_filter_expr` is for filter expressions.